sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java

```diff
@@ -0,0 +1,390 @@
+        if (pendingRequests == null) {
+            pendingRequests = new ArrayList<>();
+            pendingRequestsByIndex.put(indexName, pendingRequests);
+        }
+        pendingRequests.add(new PendingRequest<>(item, rowShardResolver.routing()));
+    }
+
+    private void sendRequestsForBulk(CompletableFuture<BitSet> bulkResultFuture, Map<ShardLocation, TReq> bulkRequests) {
+        if (bulkRequests.isEmpty()) {
+            bulkResultFuture.complete(responses);
+            return;
+        }
+        AtomicInteger numRequests = new AtomicInteger(bulkRequests.size());
+        Iterator<Map.Entry<ShardLocation, TReq>> it = bulkRequests.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<ShardLocation, TReq> entry = it.next();
+            TReq request = entry.getValue();
+            // We need to drain the bulkRequests map because when indices don't need to be created we use the global
+            // requestsByShard map. We need requestsByShard empty after a bulk is triggered for execution so it can
+            // store the requests for the next bulk.
+            it.remove();
+
+            final ShardLocation shardLocation = entry.getKey();
+            nodeJobsCounter.increment(shardLocation.nodeId);
+            ActionListener<ShardResponse> listener = new ActionListener<ShardResponse>() {
+
+                @Override
+                public void onResponse(ShardResponse shardResponse) {
+                    nodeJobsCounter.decrement(shardLocation.nodeId);
+                    processShardResponse(shardResponse);
+                    countdown();
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    nodeJobsCounter.decrement(shardLocation.nodeId);
+                    countdown();
+                }
+
+                private void countdown() {
+                    if (numRequests.decrementAndGet() == 0) {
+                        bulkResultFuture.complete(responses);
+                    }
+                }
+            };
+
+            listener = new RetryListener<>(
+                scheduler,
+                l -> {
+                    LOGGER.debug("Executing retry Listener for nodeId: {} request: {}",
+                        shardLocation.nodeId,
+                        request);
+                    requestExecutor.execute(request, l);
+                },
+                listener,
+                BACKOFF_POLICY
+            );
+            requestExecutor.execute(request, listener);
+        }
+    }
+
+    private void processShardResponse(ShardResponse shardResponse) {
+        IntArrayList itemIndices = shardResponse.itemIndices();
+        pendingItemsCount.addAndGet(-itemIndices.size());
+        List<ShardResponse.Failure> failures = shardResponse.failures();
+        synchronized (responses) {
+            for (int i = 0; i < itemIndices.size(); i++) {
```

Comment:
maybe this block could be extracted into a static method that's shared with the other Executor